### DISCUSSION TITLE
Guarantees on tree sitter AST form

### DISCUSSION THREAD

**Discussion by VonTum** [tree-sitter/tree-sitter#3206] [Q&A]
Hi, I'm looking to use tree-sitter as the parsing frontend for my compiler. 

Previous [discussions](https://github.com/tree-sitter/tree-sitter/discussions/831) have discouraged the use of tree sitter for this use case, but I think with a loose enough syntax description, the limited error information isn't an issue, which is far compensated for by tree-sitter's incredible parsing power, and ability to do local updates. 

However, in using the parser, I need to convert tree-sitter's representation to my own. It would be very tedious and perhaps redundant to do sanity checking on every single tree node, so I want to know what guarantees tree sitter provides for its generated tree. 

As an example, I've noticed that nodes with fixed size, with always either fully be ERROR, or fully have valid fields. Can I rely on this assumption? 

Ex: The grammar
```javascript
....
module: $ => seq(
    'module',
    field('name', $.identifier),
    field('block', $.block)
),
        
block: $ => seq(
    '{',
    repeat($._statement),
    '}'
),
....
``` 
For the input:
```
module a
module b {}
```
Will generate the following:
```
(source_file [6, 0] - [20, 0]
  (ERROR [6, 0] - [6, 8]
    (identifier [6, 7] - [6, 8]))
  (module [7, 0] - [7, 11]
    name: (identifier [7, 7] - [7, 8])
    block: (block [7, 9] - [7, 11])))
```
Either the module is not a module but ERROR, or it is fully valid. But for example it wouldn't be a valid `module` node, with then an ERROR `block` field. 

Are there any concrete guarantees about this behaviour? 

Another guarantee I'm interested in is if node types can occur that don't occur in the grammar. Like could the above `block` field actually contain an `identifier`?

And finally maybe some trivia, but how does tree-sitter know which grammar element is the root node? I don't think `source_file` is a builtin, like `word` is?

**Comment by sogaiu** [comment]
> how does tree-sitter know which grammar element is the root node?

I think the first thing encountered in `rules` (in `grammar.js`) is treated as the root node.  [Here](https://github.com/sogaiu/tree-sitter-clojure/blob/6e41628e9d18b19caea1cb1d72aae4ccff5bdfe9/grammar.js#L248-L252) is an example.  I got burned by this by changing the order at one point.

If that is mistaken, may be someone can correct :)

**Reply by VonTum** [reply]
I've found it strange that this is the case. I'm no Javascript expert but shouldn't key-value maps like this have an undetermined order, regardless of textual order? 

And still, is there rigorous specification on tree-sitter somewhere, because all I can find are various "introduction to tree-sitter" documents
